### PR TITLE
Implement `get_json_object_multiple_paths`

### PR DESCRIPTION
**Pull request by ttnghia**
This re-implements `get_json_object` to support extracting multiple JSON objects from the same JSON input. By processing multiple JSON paths in just one kernel call, we can improve the performance of SQL queries in which there are large numbers of calls to `get_json_object`.

The building blocks of this work are very simple:
 * Employ an entire warp of threads (32 threads) to process only one row. This eliminates thread divergence and avoids threads from blocking each other due to executing different code paths on different input.
 * Enhance memory access by using consecutive warps to process the same input rows, just different JSON paths.

Depends on:
 * https://github.com/NVIDIA/spark-rapids-jni/pull/2232

---

Benchmark on a dataset of `200'000` rows, given by a `6.7GB` parquet file, and `80` JSON paths. The data is read using parquet chunked reader with `256MB` output limit.
```
Baseline (before this work): 410 seconds.
With this work: 44 seconds.
```

### CODE REVIEW

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

**Comment by ttnghia** [comment]
build

File: src/main/cpp/src/get_json_object.cu [hunk 1]
```diff
@@ -400,10 +402,10 @@ __device__ thrust::pair<bool, cudf::size_type> evaluate_path(
     // no need to check stack is full
     // because Spark-Rapids already checked maximum length of `path_instruction`
     auto& ctx          = stack[stack_size++];
-    ctx.g              = _g;
+    ctx.g              = std::move(_g);
     ctx.path           = _path;
     ctx.dirty          = 0;
-    ctx.case_path      = _case_path;
+    ctx.case_path      = std::move(_case_path);
```
Comment (res-life) [inline src/main/cpp/src/get_json_object.cu:408]: `case_path` is a enum, need a move?

Comment (ttnghia) [reply src/main/cpp/src/get_json_object.cu:408]: Oh yes, not needed.

Comment (ttnghia) [reply src/main/cpp/src/get_json_object.cu:408]: That's my mistake. I intentionally wanted to move `_path` instead.

File: src/main/cpp/src/JSONUtilsJni.cpp [hunk 2]
```diff
@@ -67,4 +68,66 @@ JNIEXPORT jlong JNICALL Java_com_nvidia_spark_rapids_jni_JSONUtils_getJsonObject
...
+  JNI_NULL_CHECK(env, j_paths, "j_paths is null", 0);
+  JNI_NULL_CHECK(env, j_path_offsets, "j_path_offsets is null", 0);
+
+  using path_type = std::vector<std::tuple<path_instruction_type, std::string, int64_t>>;
+
+  try {
+    cudf::jni::auto_set_device(env);
+
+    auto const path_offsets = cudf::jni::native_jintArray(env, j_path_offsets).to_vector();
+    CUDF_EXPECTS(path_offsets.size() > 1, "WRONG");  // TODO
```
Comment (res-life) [inline src/main/cpp/src/JSONUtilsJni.cpp:85]: Do not forget this TODO.

File: src/main/cpp/src/get_json_object.cu [hunk 3]
```diff
@@ -789,207 +790,391 @@ __device__ thrust::pair<bool, cudf::size_type> evaluate_path(
...
+    construct_path_commands(json_paths, stream);
+
+  auto const [max_row_size, sum_row_size] =
+    thrust::transform_reduce(rmm::exec_policy(stream),
+                             thrust::make_counting_iterator(0),
+                             thrust::make_counting_iterator(input.size()),
+                             cuda::proclaim_return_type<thrust::pair<int64_t, int64_t>>(
+                               [in_offsets] __device__(auto const idx) {
+                                 auto const size = in_offsets[idx + 1] - in_offsets[idx];
+                                 return thrust::pair<int64_t, int64_t>{size, size};
```
Comment (res-life) [inline src/main/cpp/src/get_json_object.cu:1036]:
```
thrust::pair<int64_t, int64_t>{size, size};
```
return two identical size, why not just return one?

File: src/main/cpp/src/get_json_object.cu [hunk 4]
```diff
@@ -789,207 +790,391 @@ __device__ thrust::pair<bool, cudf::size_type> evaluate_path(
...
+                               }),
+                             thrust::pair<int64_t, int64_t>{0, 0},
+                             cuda::proclaim_return_type<thrust::pair<int64_t, int64_t>>(
+                               [] __device__(auto const& lhs, auto const& rhs) {
+                                 return thrust::pair<int64_t, int64_t>{
+                                   std::max(lhs.first, rhs.first), lhs.second + rhs.second};
```
Comment (res-life) [inline src/main/cpp/src/get_json_object.cu:1042]: Do not need `.first`, `.second`, refer to previous comment.

File: src/main/cpp/src/get_json_object.cu [hunk 5]
```diff
@@ -789,207 +790,391 @@ __device__ thrust::pair<bool, cudf::size_type> evaluate_path(
...
+
+    h_path_data.emplace_back(json_path_processing_data{d_json_paths[idx],
+                                                       in_offsets,
+                                                       out_stringviews.back().data(),
+                                                       scratch_buffers.back().data(),
+                                                       d_has_out_of_bound.data() + idx});
+  }
+  auto d_path_data = cudf::detail::make_device_uvector_async(
+    h_path_data, stream, rmm::mr::get_current_device_resource());
+  thrust::uninitialized_fill(
```
Comment (res-life) [inline src/main/cpp/src/get_json_object.cu:1080]: Running on GPU, because of the number of json paths is not too large, running CPU will be more fast.

**Comment by res-life** [comment]
LGTM.

**Review by res-life — APPROVED** [review]

**Comment by res-life** [comment]
We may add a case for the `out-of-bound write` scenario for multipile paths in a follow-up PR.

**Comment by ttnghia** [comment]
build

**Pull request merged**